pact/exchange.pact

```diff
@@ -0,0 +1,582 @@
+           , 'token-in: token-in
+           , 'in: amountIn
+           , 'out: (round-unit token-out (/ numerator denominator))
+           , 'idx: (+ 1 (at 'idx head))
+           , 'pair: p
+           , 'path: (drop 1 (at 'path head))
+           } ]
+         allocs)
+    )
+  )
+
+
+  (defun swap-exact-out
+    ( amountOut:decimal
+      amountInMax:decimal
+      path:[module{fungible-v2}]
+      to:string
+      guard:guard
+      deadline:time
+    )
+    (enforce-deadline deadline)
+    (enforce (>= (length path) 2) "swap-exact-in: invalid path")
+    ;; fold over tail of reverse path with dummy first value to compute inputs
+    ;; assembles allocs in forward order
+    (let*
+      ( (rpath (reverse path))
+        (path-len (length path))
+        (pz (get-pair (at 0 rpath) (at 1 rpath)))
+        (allocs
+          (fold (compute-in)
+            [ { 'token-out: (at 1 rpath)
+              , 'token-in: (at 0 rpath)
+              , 'out: 0.0
+              , 'in: amountOut
+              , 'idx: path-len
+              , 'pair: pz
+              , 'path: []
+              }]
+            (drop 1 rpath)))
+        (allocs1 ;; drop dummy at end, prepend dummy for initial transfer
+          (+ [  { 'token-out: (at 0 path)
+                , 'token-in: (at 1 path)
+                , 'out: (at 'in (at 0 allocs))
+                , 'in: 0.0
+                , 'idx: 0
+                , 'pair: (at 'pair (at 0 allocs))
+                , 'path: path
+             } ]
+             (take (- path-len 1) allocs)))
+      )
+      (enforce (<= (at 'out (at 0 allocs1)) amountInMax)
+        (format "swap-exact-in: excessive input amount {}"
+          [allocs]))
+      (with-capability (SWAPPING)
+        (swap to guard allocs1))
+    )
+  )
+
+  (defun compute-in
+    ( allocs:[object{alloc}]
+      token-in:module{fungible-v2}
+    )
+    (let*
+      ( (head:object{alloc} (at 0 allocs))
+        (token-out:module{fungible-v2} (at 'token-in head))
+        (amountOut:decimal (at 'in head))
+        (p (get-pair token-in token-out))
+        (reserveIn (reserve-for p token-in))
+        (reserveOut (reserve-for p token-out))
+        (numerator (* reserveIn amountOut))
+        (denominator (* (- reserveOut amountOut) (- 1.0 FEE)))
+      )
+      (+ [ { 'token-out: token-out
+           , 'token-in: token-in
+           , 'in: (round-unit token-in (/ numerator denominator))
+           , 'out: amountOut
+           , 'idx: (- (at 'idx head) 1)
+           , 'pair: p
+           , 'path: (+ [token-out] (at 'path head))
+           } ]
+         allocs)
+    )
+  )
+
+
+
+  (defun swap
+    ( to:string
+      guard:guard
+      allocs:[object{alloc}]
+    )
+    (require-capability (SWAPPING))
+    (let*
+      ( (head:object{alloc} (at 0 allocs))
+        (head-token:module{fungible-v2} (at 'token-out head))
+        (account (at 'account (at 'pair head)))
+        (out (at 'out head))
+      )
+      (head-token::transfer to account out)
```

Comment:
Should this be `transfer-create` so the tx will fail if the guard doesn't match the guard that was passed in?

Reply:
This is `transfer-create` now :+1: